File: apractice/sort/QuickSort.java
import java.util.Arrays;

// package sorts;

public class QuickSort {
    public static void quickSort(int[] arr, int n) {
        quickSortInternally(arr, 0, n - 1);
    }

    /**
     * 递归函数
     */
    public static void quickSortInternally(int[] arr, int leftmark, int rightmark) {
        if (p >= rightmark) return;

        int pivot = partition(arr, leftmark, rightmark);  // 获得区分点
        quickSortInternally(arr, leftmark, pivot-1);      // left half
        quickSortInternally(arr, pivot+1, rightmark);     // right half
    }

    private static int partition(int[] arr, int leftmark, int rightmark) {
        int pivot = arr[r];                            // pivot
        int i = leftmark;                              // i leftmark
        for (int j = leftmark; j < rightmark; ++j) {   // j 遍历 index
            if (arr[j] < pivot) {
                if (i == j) {
                    ++i;                               // 当前值小于 pivot, leftmart 右移1个位置
                } else {
                    int tmp = arr[i];                  // 当前值不小于 pivot, 交换
                    arr[i++] = arr[j];
                    arr[j] = tmp;
                }
            }
        }

        int tmp = arr[i];                              // 把基准值放到正确的位置
        arr[i] = arr[r];
        arr[r] = tmp;

        System.out.println("i=" + i + "     " + Arrays.toString(arr));
        return i;
    }

    public static void main(String[] args) {
        int[] arr1 = new int[]{3, 2, 52, 43, 32, 56, 93, 29, 19};
        System.out.println("Origin: " + Arrays.toString(arr1));
        quickSort(arr1, arr1.length);
        System.out.println("Sorted: " + Arrays.toString(arr1));
    }
}

File: apractice/sort/quickSort.py
def quickSort(alist):
    quickSortHelper(alist,0,len(alist)-1)


def quickSortHelper(alist, first, last):
    if first >= last:
        return

    splitpoint = partition(alist, first, last)
    quickSortHelper(alist, first, splitpoint-1)
    quickSortHelper(alist, splitpoint+1, last)


def partition(alist, first, last):
    pivotvalue = alist[last]

    leftmark = first
    rightmark = last - 1

    done = False
    while not done:
        while leftmark <= rightmark and alist[leftmark] <= pivotvalue:
            leftmark = leftmark + 1

        while leftmark <= rightmark and alist[rightmark] >= pivotvalue:
            rightmark = rightmark -1

        if rightmark < leftmark:
            done = True
        else:
            temp = alist[leftmark]
            alist[leftmark] = alist[rightmark]
            alist[rightmark] = temp

    temp = alist[last]
    alist[last] = alist[leftmark]
    alist[leftmark] = temp

    return leftmark

alist = [54,26,93,17,77,31,44,55,20,44]
quickSort(alist)
print(alist)
